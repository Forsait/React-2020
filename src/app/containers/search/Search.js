import React from 'react';

import { connect } from 'react-redux';
import { homeChange } from '../../actions/home';

import styles from './Search.module.scss';
import Header from '../../components/common/Header';
import Westside from '../../components/common/Westside';
import SearchForm from '../../components/searchForm/SearchForm';
import Radiobuttons from '../../components/common/Radiobuttons';

export class Search extends React.Component {
  constructor(props){
    super(props);
    this.submitForm = this.submitForm.bind(this);
    this.sortByChange = this.sortByChange.bind(this);
    this.searchByChange = this.searchByChange.bind(this);

    this.state = {
      searchBy: 'title',
      sortBy: 'release_date',
      query: '',
    };
  }

  componentDidMount() {
    this.submit();
  }

  submit(stateNexValue = {}) {
    this.props.submit({
      searchBy: this.state.searchBy,
      query: this.state.query,
      sortBy: this.state.sortBy,
      ...stateNexValue
    });
  }

  submitForm(query) {
    this.setState({
      query,
    });
    this.submit({query});
  }

  sortByChange(sortBy) {
    this.setState({
      sortBy,
    });
    this.submit({sortBy});
  }

  searchByChange(searchBy) {
    this.setState({
      searchBy,
    });
    this.submit({searchBy});
  }

  render() {
    let resTotal = '';
    if (this.state.isSearchSubmit) {
      resTotal = `${+this.props.resultTotal} movie found`;
    }

    return (
      <>
        <div className={styles.search}>
          <div className={styles.back_image}></div>
          <Header className={styles.brand} />

          <div className={styles.form_container}>
            <h1>FIND YOUR MOVIE</h1>
            <SearchForm submitForm={this.submitForm} />
            <Radiobuttons 
              className={styles.t_1} 
              title={this.props.searchByOptions.title} 
              list={this.props.searchByOptions.list} 
              change={this.searchByChange}/>
          </div>
        </div>
        <Westside className={styles.result_container}>
          <div className={styles.q_1}>{resTotal}</div>
          <Radiobuttons 
            className={styles.r_1}
            title={this.props.sortByOptions.title}
            list={this.props.sortByOptions.list}
            change={this.sortByChange}/>
        </Westside>
      </>
    )
  }
}

function mapStateToProps(state) {
  const { moviesData, searchByOptions, sortByOptions } = state.home;
  return {
    movies: moviesData,
    searchByOptions,
    sortByOptions,
  };
}

const mapDispatchToProps = dispatch => {
  return {
    submit: conf => dispatch(homeChange(conf))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Search);