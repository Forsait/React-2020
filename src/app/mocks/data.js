export const film = {
  budget: 0,
  genres: ['Action', 'Adventure', 'Drama', 'Science Fiction', 'Thriller'],
  id: 351286,
  overview: 'A volcanic eruption threatens',
  poster_path: 'https://image.tmdb.org/t/p/w500/ln6d5Okr6VK5vfQVobJTiYxeD0l.jpg',
  release_date: '2018-06-01',
  revenue: 0,
  runtime: null,
  tagline: 'Life finds a way',
  title: 'Jurassic World: Fallen Kingdom',
  vote_average: 0,
  vote_count: 28,
};

export const moviesData = {
  data: new Array(6).fill(film),
  limit: 0,
  offset: 0,
  total: 0,
};

export const radioOptions = {
  title: 'Radio title',
  list: [
    {
      key: 'key 1',
      title: 'title 1',
    },
    {
      key: 'key 2',
      title: 'title 2',
    },
  ],
};
