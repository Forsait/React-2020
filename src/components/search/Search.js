import React from 'react';

import styles from './Search.module.scss';
import Brand from '../common/Brand';
import SearchForm from './SearchForm';
import Radiobuttons from '../common/Radiobuttons';

const options = {
  title: 'Sort by',
  list: [
    {
      key: 'release_date',
      title: 'release date'
    },
    {
      key: 'vote_average',
      title: 'rating'
    }
  ]
};

const checkOptions = {
  title: 'Search by',
  list: [
    {
      key: 'title',
      title: 'title'
    },
    {
      key: 'genres',
      title: 'genre'
    }
  ]
}


export default class Search extends React.Component {
  constructor(props){
    super(props);
    this.submitForm = this.submitForm.bind(this);
    this.sortByChange = this.sortByChange.bind(this);
    this.searchByChange = this.searchByChange.bind(this);

    this.state = {
      isSearchSubmit: false,
      searchBy: 'title',
      sortBy: 'release_date',
      query: '',
    }
  }

  submit(stateNexValue) {
    this.props.submit({
      searchBy: this.state.searchBy,
      query: this.state.query,
      sortBy: this.state.sortBy,
      ...stateNexValue
    });
  }

  submitForm(query) {
    this.setState({
      query,
      isSearchSubmit: true
    });
    this.submit({query});
  }

  sortByChange(sortBy) {
    this.setState({
      sortBy,
      isSearchSubmit: true
    });
    this.submit({sortBy});
  }

  searchByChange(searchBy) {
    this.setState({
      searchBy,
      isSearchSubmit: true
    });
    this.submit({searchBy});
  }

  render() {
    let resTotal = '';
    if (this.state.isSearchSubmit && this.props.resultTotal > 0) {
      resTotal = `${this.props.resultTotal} movie found`
    }

    return (
      <>
        <div className={styles.search}>
          <div className={styles.back_image}></div>
          <Brand className={styles.brand} />

          <div className={styles.form_container}>
            <h1>FIND YOUR MOVIE</h1>
            <SearchForm submitForm={this.submitForm} />
            <Radiobuttons 
              className={styles.t_1} 
              title={checkOptions.title} 
              list={checkOptions.list} 
              change={this.searchByChange}/>
          </div>
        </div>
        <div className={`${styles.result_container} westside`}>
          <div className={styles.q_1}>{resTotal}</div>
          <Radiobuttons 
            className={styles.r_1}
            title={options.title}
            list={options.list}
            change={this.sortByChange}/>
        </div>
      </>
    )
  }
}