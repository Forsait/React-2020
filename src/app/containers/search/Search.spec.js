import React from 'react';
import { shallow } from 'enzyme';

import { Search } from './Search';
import { radioOptions, moviesData } from '../../mocks/data';

describe('Search component', () => {

  it('should be render correctly', () => {
    const submit = jest.fn();
    Search.prototype.submit = submit;
    const com = shallow(<Search movies={moviesData} searchByOptions={radioOptions} sortByOptions={radioOptions} />);
    expect(com).toMatchSnapshot()
  });

  it('Should run submit method after submit form', () => {
    const submit = jest.fn();
    Search.prototype.submit = submit;
    const com = shallow(<Search movies={moviesData} searchByOptions={radioOptions} sortByOptions={radioOptions} />);
    com.instance().submitForm('value');
    expect(submit).toHaveBeenCalledWith({query: 'value'})
  });

  it('Should run submit method after sortBy change', () => {
    const submit = jest.fn();
    Search.prototype.submit = submit;
    const com = shallow(<Search movies={moviesData} searchByOptions={radioOptions} sortByOptions={radioOptions} />);
    com.instance().sortByChange('value');
    expect(submit).toHaveBeenCalledWith({sortBy: 'value'})
  });

  it('Should run submit method after searchBy change', () => {
    const submit = jest.fn();
    Search.prototype.submit = submit;
    const com = shallow(<Search movies={moviesData} searchByOptions={radioOptions} sortByOptions={radioOptions} />);
    com.instance().searchByChange('value');
    expect(submit).toHaveBeenCalledWith({searchBy: 'value'})
  });

});
