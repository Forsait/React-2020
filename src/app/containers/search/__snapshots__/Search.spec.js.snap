// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component should be render correctly 1`] = `
<Fragment>
  <div>
    <div />
    <Header />
    <div>
      <h1>
        FIND YOUR MOVIE
      </h1>
      <SearchForm
        submitForm={[Function]}
      />
      <Radiobuttons
        change={[Function]}
        list={
          Array [
            Object {
              "key": "key 1",
              "title": "title 1",
            },
            Object {
              "key": "key 2",
              "title": "title 2",
            },
          ]
        }
        title="Radio title"
      />
    </div>
  </div>
  <Westside>
    <div />
    <Radiobuttons
      change={[Function]}
      list={
        Array [
          Object {
            "key": "key 1",
            "title": "title 1",
          },
          Object {
            "key": "key 2",
            "title": "title 2",
          },
        ]
      }
      title="Radio title"
    />
  </Westside>
</Fragment>
`;
